src/common/framework/params_builder.ts

```diff
@@ -158,9 +160,17 @@ export class CaseParamsBuilder<CaseP extends {}>
     expander: (_: Merged<{}, CaseP>) => Iterable<NewPValue>
   ): CaseParamsBuilder<Merged<CaseP, { [name in NewPKey]: NewPValue }>> {
     return this.expandWithParams(function* (p) {
+      const seenValues = new Set();
       for (const value of expander(p)) {
         // TypeScript doesn't know here that NewPKey is always a single literal string type.
-        yield { [key]: value } as { [name in NewPKey]: NewPValue };
+        const entry = { [key]: value } as { readonly [name in NewPKey]: NewPValue };
+        assert(
+          !seenValues.has(value),
+          `Duplicate value for param "${key}": ${stringifyPublicParams({ ...p, ...entry })}`
+        );
```

Comment:
Oh, this checks for duplicates on every iteration. I was going to fix that and only do it once.